### PR TITLE
[DebugInfo][RemoveDIs] Handle dbg.declares in SelectionDAGISel

### PR DESCRIPTION
**Pull request by OCHyams**
This is a boring mechanical update to support DPValues that look like dbg.declares in SelectionDAG.

---

The only question I have is, how many tests should get upgraded with `--try-experimental-debuginfo-iterators`? The set that's included in this patch is arbitrary-ish -- I tried to pick ones that looked like they were probably testing the SelectionDAGBuilder specifically, but a bunch more snuck in too. Should I add it to all MIR tests that include a dbg.declare?

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp [hunk 1]
```diff
@@ -1148,6 +1148,60 @@ SDValue SelectionDAGBuilder::getControlRoot() {
...
+
+  bool IsParameter = Variable->isParameter() || isa<Argument>(Address);
+
+  SDValue &N = NodeMap[Address];
+  if (!N.getNode() && isa<Argument>(Address))
+    // Check unused arguments map.
+    N = UnusedArgNodeMap[Address];
+  SDDbgValue *SDV;
+  if (N.getNode()) {
+    if (const BitCastInst *BCI = dyn_cast<BitCastInst>(Address))
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:1174]: Do you really need this with opaque pointers? stripPointerCasts?

Comment (OCHyams) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:1174]: I've just moved this blob of existing code from `SelectionDAGBuilder::visitIntrinsicCall` into its own function.

I could change this in a separate patch?

File: llvm/include/llvm/CodeGen/FunctionLoweringInfo.h [hunk 2]
```diff
@@ -191,6 +191,7 @@ class FunctionLoweringInfo {
   /// Collection of dbg.declare instructions handled after argument
   /// lowering and before ISel proper.
   SmallPtrSet<const DbgDeclareInst *, 8> PreprocessedDbgDeclares;
+  SmallPtrSet<const DPValue *, 8> PreprocessedDPVs;
```
Comment (jmorse) [inline llvm/include/llvm/CodeGen/FunctionLoweringInfo.h:194]: What do you think about DPVDeclares, to distinguish this from "normal" DPValues? (Alas, naming things is one of the two difficult parts of CS).

**Review by jmorse — APPROVED** [review]
LGTM; Matt raises a good point about the bitcast, but as this is just a refactor + adding two new calls to handle dbg.declares stored in DPValues, I think it's best treated elsewhere. We've already got some tickets hanging around about how the handling of debug-info for arguments in SelectionDAG is a bit of a hack.

File: llvm/include/llvm/CodeGen/FunctionLoweringInfo.h (hunk 2 above)
Comment (OCHyams) [reply llvm/include/llvm/CodeGen/FunctionLoweringInfo.h:194]: Do you have a preference between PreprocessedDPVDeclares and PreprocessedDPDeclares (no V, as in in DPV, in the second one)? The former, DPVDeclares, might be more grep-friendly?

Comment (jmorse) [reply llvm/include/llvm/CodeGen/FunctionLoweringInfo.h:194]: Let's go with the former, DPV indicating "debug-info things are happening now" is an alright (if not perfect) thing to maintain, "DP" by itself is open to more interpretation.

**Pull request merged**